java-squid/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValueRelation.java

```diff
@@ -0,0 +1,270 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.se.symbolicvalues;
+
+import com.google.common.collect.ImmutableList;
+import org.sonar.java.se.ConstraintManager.BooleanConstraint;
+import org.sonar.java.se.ProgramState;
+
+import javax.annotation.CheckForNull;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeSet;
+
+public abstract class SymbolicValueRelation extends BinarySymbolicValue {
+
+  static final String TRANSITIVE_RELATIONS_EXCEEDED = "Number of transitive relations exceeded!";
+  private static final int VIRTUAL = -1;
+
+  static class RelationComparator implements Comparator<SymbolicValueRelation> {
+
+    @Override
+    public int compare(SymbolicValueRelation rel1, SymbolicValueRelation rel2) {
+
+      final int[] key1 = rel1.key();
+      final int[] key2 = rel2.key();
+      int answer = Integer.compare(key1[0], key2[0]);
+      if (answer == 0) {
+        answer = Integer.compare(key1[1], key2[1]);
+        if (answer == 0) {
+          answer = rel1.getClass().getName().compareTo(rel2.getClass().getName());
+        }
+      }
+      return answer;
+    }
+
+  }
+
+  public SymbolicValueRelation(int id) {
+    super(id);
+  }
+
+  protected SymbolicValueRelation(SymbolicValue v1, SymbolicValue v2) {
+    super(VIRTUAL);
+    computedFrom(ImmutableList.of(v2, v1));
+  }
+
+  @Override
+  public BooleanConstraint shouldNotInverse() {
+    return BooleanConstraint.FALSE;
+  }
+
+  @Override
+  public boolean isVirtual() {
+    return id() == VIRTUAL;
+  }
+
+  @Override
+  protected List<ProgramState> copyConstraint(SymbolicValue from, SymbolicValue to, ProgramState programState, BooleanConstraint booleanConstraint) {
+    ProgramState newState = programState;
+    if (programState.canReach(from) || programState.canReach(to)) {
+      newState = programState.addConstraint(this, booleanConstraint);
+    }
+    return super.copyConstraint(from, to, newState, booleanConstraint);
+  }
+
+  private int[] key() {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Move this method into "RelationComparator". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS3398)